Python/chapter01/1.9 - String Rotation/miguel_1.9_sol.py

```diff
@@ -0,0 +1,83 @@
+"""
+Python version 3.7.0
+1.9 - String Rotation
+Assume you have a method isSubstring which
+checks if one word is a substring
+of another.  Given two strings, s1 and s2,
+write code to check if s2 is a
+rotation of s1 using only one call to isSubstring
+(e.g. 'waterbottle' is a rotation of 'erbottlewat')
+"""
+import unittest
+
+
+def is_substring(sub: str, s: str) -> bool:
+    """
+    is_substring checks if 'sub' is a substring of 's'
+    :param sub: the substring
+    :param s: the alleged full string
+    :return: true if sub is a substring of s, false otherwise
+    """
+    return sub in s
+
+
+def string_rotation(s1: str, s2: str) -> bool:
+    """
+    Given two strings, string_rotation will check if s2 is a rotation of s1
+    using only one call to isSubstring.
+    Assume s2 is a rotation of s1.  s1 is a rearrangement of s2.
+    Choosing an index where we rotate to be any index,
+    let s1 and s2 consist of two parts, a and b.
+    The original string will be s2 = ab.
+    The rotated string will be s1 = ba.
+    Since we are using substring, we want s2 = ab to appear in
+    a larger string ..ab..
+    This is guaranteed to happen if we have s1 + s1 = baba
+    because ab shows up in there.
+    Runtime: worst case: O(n^2) - substring found at end of full string
+    Space Complexity:  O(1)
```

Comment:
argh, i missed this one. you are now constructing a brand new string, so this is no longer O(1) (although with some clever data structures like Cords it could have been).

Reply:
Going to update the comment now.